### PR TITLE
[SPARK-40817][K8S] `spark.files` should preserve remote files

### PR DESCRIPTION
**Pull request by antonipp**
### What changes were proposed in this pull request?
You can find a detailed description of the issue and an example reproduction on the Jira card: https://issues.apache.org/jira/browse/SPARK-40817

The idea for this fix is to update the logic which uploads user-specified files (via `spark.jars`, `spark.files`, etc) to `spark.kubernetes.file.upload.path`. After uploading local files, it used to overwrite the initial list of URIs passed by the user and it would thus erase all remote URIs which were specified there.

Small example of this behaviour: 
1. User set the value of `spark.jars` to `s3a://some-bucket/my-application.jar,/tmp/some-local-jar.jar` when running `spark-submit` in cluster mode
2. `BasicDriverFeatureStep.getAdditionalPodSystemProperties()` gets called at one point while running `spark-submit`
3. This function would set `spark.jars` to a new value of `${SPARK_KUBERNETES_UPLOAD_PATH}/spark-upload-${RANDOM_STRING}/some-local-jar.jar`. Note that `s3a://some-bucket/my-application.jar` has been discarded.

With the logic proposed in this PR, the new value of `spark.jars` would be `s3a://some-bucket/my-application.jar,${SPARK_KUBERNETES_UPLOAD_PATH}/spark-upload-${RANDOM_STRING}/some-local-jar.jar`, so in other words we are making sure that remote URIs are no longer discarded.

### Why are the changes needed?
We encountered this issue in production when trying to launch Spark on Kubernetes jobs in cluster mode with a mix of local and remote dependencies.

### Does this PR introduce _any_ user-facing change?
Yes, see description of the new behaviour above.

### How was this patch tested?
- Added a unit test for the new behaviour
- Added an integration test for the new behaviour
- Tried this patch in our Kubernetes environment with `SparkPi`:
```
spark-submit \
  --master k8s://https://$KUBERNETES_API_SERVER_URL:443 \
  --deploy-mode cluster \
  --name=spark-submit-test \
  --class org.apache.spark.examples.SparkPi \
  --conf spark.jars=/opt/my-local-jar.jar,s3a://$BUCKET_NAME/my-remote-jar.jar \
  --conf spark.kubernetes.file.upload.path=s3a://$BUCKET_NAME/my-upload-path/ \
  [...]
  /opt/spark/examples/jars/spark-examples_2.12-3.1.3.jar
```
Before applying the patch, `s3a://$BUCKET_NAME/my-remote-jar.jar` was discarded from the final value of `spark.jars`. After applying the patch and launching the job again, I confirmed that `s3a://$BUCKET_NAME/my-remote-jar.jar` was no longer discarded by looking at the Spark config for the running job.

### CODE REVIEW

**Comment by antonipp** [comment]
Note that the only test which is failing is `pyspark-pandas-slow` which is completely unrelated to my changes. The error is: 
```
ERROR: No matching distribution found for py4j==0.10.9.7
```
which looks like a dependency issue. I didn't dig much further than that. 
All other tests pass.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 1]
```diff
@@ -1609,6 +1609,16 @@ class TestFileSystem extends org.apache.hadoop.fs.LocalFileSystem {
   }
 
   override def open(path: Path): FSDataInputStream = super.open(local(path))
+
+  // No-op methods
+
+  override def copyFromLocalFile(
```
Comment (dcoliversun) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: +1 on the idea. Why we need override methods `copyFromLocalFile` and `mkdirs` here?

Comment (antonipp) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: This was needed for the unit test to work. I was looking for a way to mock the [KubernetesUtils.uploadFileUri()](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L310-L335) method which is used to actually upload the local files to a remote path specified in `spark.kubernetes.file.upload.path`. `KubernetesUtils` is an `object` which are notoriously hard to mock and there was a bit too much refactoring required to make this method mockable.

So I decided to go with another approach: use a fake Hadoop FS so that the operations performed in the `uploadFileUri()` method become no-ops. I found this `TestFileSystem` class which was used in other tests for the same purpose so I simply added the methods which were called inside `uploadFileUri()`: `mkdirs` [here](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L320) and `copyFromLocalFile` [here](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L347). I then made sure that this FS implementation was used by the test by specifying it here: https://github.com/antonipp/spark/blob/80416f6a6ff42051cd38dfe32df2d0c7cd0741f4/resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala#L358

Comment (dcoliversun) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: nit: Each level of indentation is 2 spaces in apache spark.  https://docs.scala-lang.org/style/indentation.html  
BTW, if you want to test with `KubernetesUtils`, you could add integration test with MINIO in `DepsTestsSuite`.

Comment (antonipp) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: Fixed the indentation in a new commit.

I also spent some time trying to make integration tests work on my machine but encountered a blocking issue with my local setup, I opened a ticket here: https://issues.apache.org/jira/browse/SPARK-40954 
Could we merge this PR as it is? I can spend more time on fixing my integration test setup + writing a new test some time later but I think the unit test I added already gives us pretty good coverage.

Comment (dcoliversun) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: Thanks. I have left comments on [SPARK-40954](https://issues.apache.org/jira/browse/SPARK-40954). I'm not sure if this issue is only on the M1, I haven't reproduced the issue on my Intel Mac. Maybe try using `hyperkit` first to see if the issue recurs. This PR needs others to review, let me @dongjoon-hyun @holdenk

Comment (antonipp) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1615]: Well, it took me a a bit of time to get to it but I finally have a working local setup and I wrote an integration test for this PR. I added it in https://github.com/apache/spark/pull/38376/commits/cbda5403b00f4e66ad4651ceda7786de1f5e1e1a

It worked for me locally with:
```bash
mvn integration-test -am -pl :spark-kubernetes-integration-tests_2.12 \
                        -Pkubernetes -Pkubernetes-integration-tests -Phadoop-3 \
                        -Dspark.kubernetes.test.sparkTgz=/path/to/the/tgz/built/from/this/branch \
                        -Dspark.kubernetes.test.deployMode=minikube \
                        -Dspark.kubernetes.test.imageRepo=docker.io/kubespark \
                        -Dspark.kubernetes.test.namespace=spark \
                        -Dspark.kubernetes.test.serviceAccountName=spark
```

@dongjoon-hyun @holdenk, would appreciate your reviews as well 🙏

**Review by holdenk — APPROVED** [review]
LGTM thanks for the fix. If no one has any concerns by EOW I'll try and merge pending CI passing.

**Comment by antonipp** [comment]
@holdenk thank you for the review! 
I realised that the integration test I wrote didn't work properly in the CI. I believe this was due to the fact the SparkPi job I used in the test was finishing way too quickly and Executor pods were cleaned way too quickly as well, so the `doBasicExecutorPodCheck` wasn't succeeding. I believe I fixed it with https://github.com/apache/spark/pull/38376/commits/3facb0d4318f2d8c9f7df689422434ad36aac3f6, all tests are green now! 
I think we should be good to merge?

**Comment by antonipp** [comment]
@holdenk @dcoliversun gentle reminder for this one 😄

**Comment by bjornjorgensen** [comment]
@dongjoon-hyun

**Comment by bjornjorgensen** [comment]
@dongjoon-hyun can you have a look at this one? It seams to be a pr we will have in 3.4

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 2]
```diff
@@ -1609,6 +1609,16 @@ class TestFileSystem extends org.apache.hadoop.fs.LocalFileSystem {
...
+    delSrc: Boolean,
+    overwrite: Boolean,
+    src: Path,
+    dst: Path): Unit = {}
+
+  override def mkdirs(path: Path): Boolean = true
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:1621]: Please revert this change and make a new test file system in `BasicDriverFeatureStepSuite.scala` and use it.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala [hunk 3]
```diff
@@ -344,6 +345,33 @@ class BasicDriverFeatureStepSuite extends SparkFunSuite {
...
+    val FILE_UPLOAD_PATH = "s3a://some-bucket/upload-path"
+    val REMOTE_JAR_URI = "s3a://some-bucket/my-application.jar"
+    val LOCAL_JAR_URI = "/tmp/some-local-jar.jar"
+
+    val sparkConf = new SparkConf()
+      .set(CONTAINER_IMAGE, "spark-driver:latest")
+      .set(JARS, Seq(REMOTE_JAR_URI, LOCAL_JAR_URI))
+      .set(KUBERNETES_FILE_UPLOAD_PATH, FILE_UPLOAD_PATH)
+      // Instead of using the real S3A Hadoop driver, use a fake local one
+      .set("spark.hadoop.fs.s3a.impl", classOf[TestFileSystem].getCanonicalName)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala:358]: This one is related to my previous comment, https://github.com/apache/spark/pull/38376/files#r1069907920 .

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala [hunk 4]
```diff
@@ -344,6 +345,33 @@ class BasicDriverFeatureStepSuite extends SparkFunSuite {
...
+    val sparkJars = featureStep.getAdditionalPodSystemProperties()(JARS.key).split(",")
+
+    // Both the remote and the local JAR should be there
+    assert(sparkJars.size == 2)
+    // The remote JAR path should have been left untouched
+    assert(sparkJars.contains(REMOTE_JAR_URI))
+    // The local JAR should have been uploaded to spark.kubernetes.file.upload.path
+    assert(!sparkJars.contains(LOCAL_JAR_URI))
+    assert(sparkJars.exists(path =>
+      path.startsWith(FILE_UPLOAD_PATH) && path.endsWith("some-local-jar.jar")))
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala:372]: Thank you for adding this test case.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 5]
```diff
@@ -239,6 +239,41 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
     }
   }
 
+  test(
+    "SPARK-40817: Check that remote files do not get discarded in spark.files",
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:243]:
```scala
- test(
-   "SPARK-40817: Check that remote files do not get discarded in spark.files",
-   k8sTestTag,
-   MinikubeTag) {
+ test("SPARK-40817: spark.files should preserve remote files", k8sTestTag, MinikubeTag) {
```

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 6]
```diff
@@ -239,6 +239,41 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
     }
   }
 
+  test(
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:242]: Shall we move this to line 169 because this test case is similar with `test("Launcher client dependencies", k8sTestTag, MinikubeTag)`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 7]
```diff
@@ -239,6 +239,41 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+
+      // Create a remote file on S3
+      val remoteFileKey = "some-path/some-remote-file.txt"
+      createS3Object(remoteFileKey, "Some Content")
+      val remoteFileFullPath = s"s3a://${BUCKET}/${remoteFileKey}"
+
+      // Put both file paths in spark.files
+      sparkAppConf.set("spark.files", s"$HOST_PATH/$localFileName,${remoteFileFullPath}")
+      // Allows to properly read executor logs once the job is finished
+      sparkAppConf.set("spark.kubernetes.executor.deleteOnTermination", "false")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:258]: Do we need this for this automated unit test case? I'm wondering if this is for human.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 8]
```diff
@@ -239,6 +239,41 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+      // Run SparkPi and make sure that both files have been properly downloaded on running pods
+      val examplesJar = Utils.getTestFileAbsolutePath(getExamplesJarName(), sparkHomeDir)
+      runSparkApplicationAndVerifyCompletion(
+        appResource = examplesJar,
+        mainClass = SPARK_PI_MAIN_CLASS,
+        appArgs = Array(),
+        expectedDriverLogOnCompletion = Seq("Pi is roughly 3"),
+        // We can check whether the Executor pod has successfully
+        // downloaded both the local and the remote file
+        expectedExecutorLogOnCompletion = Seq(localFileName, "some-remote-file.txt"),
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:269]: Let's define `val remoteFileName` at line 248 after `val localFileName` and re-use it in this test case (line 251 and here).

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 9]
```diff
@@ -259,20 +294,46 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+      secretKey: String = SECRET_KEY): AmazonS3Client = {
+    val credentials = new BasicAWSCredentials(accessKey, secretKey)
+    val s3client = new AmazonS3Client(credentials)
+    s3client.setEndpoint(s3Endpoint)
+    s3client
+  }
+
+  private def createS3Bucket(
+      s3Endpoint: String = getServiceUrl(svcName),
+      bucket: String = BUCKET): Unit = {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:309]: Does this PR need to change this function signature when we use always `BUCKET` during the tests? If not, please move this out from this bug-fix PR. You can do this kind of enhancement separately in a different PR.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 10]
```diff
@@ -323,7 +384,7 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
     try {
       setupMinioStorage()
       val minioUrlStr = getServiceUrl(svcName)
-      createS3Bucket(ACCESS_KEY, SECRET_KEY, minioUrlStr)
+      createS3Bucket(minioUrlStr)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:387]: This is related to the previous comment, https://github.com/apache/spark/pull/38376/files#r1069933779 . Please try to reduce the patch size by focusing on the required change .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I finished my first-round review and left a few comments. Thank you for pinging me, @antonipp , @dcoliversun , @bjornjorgensen . Since this is a bug fix, this is not blocked by Apache Spark 3.4 Feature Feeze schedule. After the full review, we can backport this to the old branches.

**Comment by bjornjorgensen** [comment]
@antonipp Is this related to https://github.com/apache/spark/pull/39488 ?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala [hunk 11]
```diff
@@ -168,27 +168,27 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
       MEMORY_OVERHEAD_FACTOR.key -> defaultOverheadFactor.toString)
     // try upload local, resolvable files to a hadoop compatible file system
     Seq(JARS, FILES, ARCHIVES, SUBMIT_PYTHON_FILES).foreach { key =>
-      val uris = conf.get(key).filter(uri => KubernetesUtils.isLocalAndResolvable(uri))
+      val (localUris, remoteUris) =
+        conf.get(key).partition(uri => KubernetesUtils.isLocalAndResolvable(uri))
       val value = {
         if (key == ARCHIVES) {
```
Comment (HyukjinKwon) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala:174]: In case of archives, it should also resolve the URL properly for `remoteUris` too in order to remove fragments (e.g., `file:///tmp/tmp4542734800151332666.txt.tar.gz#test_tar_gz`). Otherwise, it would fail to upload the file.

**Comment by HyukjinKwon** [comment]
> @antonipp Is this related to https://github.com/apache/spark/pull/39488 ?

I don't think this is related. Maybe `isLocalDependency` itself is related because it assumes that `null` in scheme is a local file (which is wrong I guess? because it assumes the default URL scheme is a file) but this PR itself isn't related to that.

**Review by antonipp — COMMENTED** [review]
Thank you for the reviews @dongjoon-hyun and @HyukjinKwon, I believe I've addressed your comments, let me know what you think!

@bjornjorgensen I agree with @HyukjinKwon that https://github.com/apache/spark/pull/39488 doesn't seem directly related

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala [hunk 12]
```diff
@@ -353,3 +381,16 @@ class BasicDriverFeatureStepSuite extends SparkFunSuite {
 
   private def amountAndFormat(quantity: Quantity): String = quantity.getAmount + quantity.getFormat
 }
+
+/**
+ * No-op Hadoop FileSystem
+ */
+private class TestFileSystem extends org.apache.hadoop.fs.LocalFileSystem {
+  override def copyFromLocalFile(
+                                  delSrc: Boolean,
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala:390]: Indentation?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala [hunk 13]
```diff
@@ -353,3 +381,16 @@ class BasicDriverFeatureStepSuite extends SparkFunSuite {
 
   private def amountAndFormat(quantity: Quantity): String = quantity.getAmount + quantity.getFormat
 }
+
+/**
+ * No-op Hadoop FileSystem
+ */
+private class TestFileSystem extends org.apache.hadoop.fs.LocalFileSystem {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala:388]: Please import `org.apache.hadoop.fs.LocalFileSystem`.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for update. From my side, it looks good to me.
For the archive stuff, we need @HyukjinKwon 's confirmation.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 11 above)
Comment (HyukjinKwon) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala:174]: 👌

**Review by HyukjinKwon — APPROVED** [review]
LGTM2

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon !

**Comment by dongjoon-hyun** [comment]
Merged to master. Could you make a backporting PR, @antonipp ?

**Comment by antonipp** [comment]
Thank you for the reviews and for the merge! 
I am not 100% sure what is the backport process but I opened 2 PRs (for 3.3 and 3.2) since I believe both are still supported based on the [versioning policy](https://spark.apache.org/versioning-policy.html).
- 3.3: https://github.com/apache/spark/pull/39669
- 3.2: https://github.com/apache/spark/pull/39670

**Comment by dongjoon-hyun** [comment]
Perfect, @antonipp !

**Comment by dongjoon-hyun** [comment]
All PRs (master/3.3/3.2) are merged.

I added to you the Apache Spark contributor group and assign SPARK-40817 to you.

Welcome to the Apache Spark community, @antonipp !
